datafusion/core/src/datasource/object_store.rs

```diff
@@ -81,10 +81,19 @@ impl std::fmt::Display for ObjectStoreUrl {
     }
 }
 
+/// Object store provider can detector an object store based on the url
+pub trait ObjectStoreProvider: Send + Sync + 'static {
+    /// Detector a suitable object store based on its url if possible
+    /// Return the key and object store
+    fn get_by_url(&self, url: &Url) -> Option<(String, Arc<dyn ObjectStore>)>;
```

Comment:
I'm not sure how this returning of a `String` is intended to work, as it will always be looked up in the hashmap based on `&url[url::Position::BeforeScheme..url::Position::BeforePath];`

Reply:
You are right. With the new logic of get_by_url, we can just use the key as the parameter and the returned value without the String key.